Sources/NIO/Codec.swift

```diff
@@ -790,3 +790,270 @@ extension MessageToByteHandler {
         }
     }
 }
+
+
+/// A simplified version of `ByteToMessageDecoder` that can generate zero or one messages for each invocation of `decode` or `decodeLast`.
+/// Having `decode` and `decodeLast` return an optional message avoids re-entrancy problems, since the functions relinquish exclusive access
+/// to the `ByteBuffer` when returning. This allows for greatly simplified (and more performant) processing.
+/// Many `ByteToMessageDecoder`'s can trivially be translated to `NIOSingleStepByteToMessageDecoder`'s.
+public protocol NIOSingleStepByteToMessageDecoder: ByteToMessageDecoder {
+    /// The type of the messages this `NIOSingleStepByteToMessageDecoder` decodes to.
+    associatedtype InboundOut
```

Comment:
It seems that this isn't possible, at least without compiler warnings. I raised [a ticket](https://bugs.swift.org/browse/SR-11868) since I think this is a swift bug.